src/Surfnet/AzureMfa/Application/Service/AzureMfaService.php

```diff
@@ -0,0 +1,177 @@
+<?php declare(strict_types=1);
+
+/**
+ * Copyright 2019 SURFnet B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\AzureMfa\Application\Service;
+
+use Exception;
+use RobRichards\XMLSecLibs\XMLSecurityKey;
+use SAML2\Assertion;
+use SAML2\Certificate\PrivateKeyLoader;
+use SAML2\Configuration\PrivateKey;
+use Surfnet\SamlBundle\Entity\IdentityProvider;
+use Surfnet\SamlBundle\Entity\ServiceProvider;
+use Surfnet\SamlBundle\Http\PostBinding;
+use Surfnet\SamlBundle\SAML2\AuthnRequest;
+use Surfnet\SamlBundle\SAML2\AuthnRequestFactory;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+
+class AzureMfaService
+{
+
+    /**
+     * @var string
+     */
+    private $publicKey;
+    /**
+     * @var string
+     */
+    private $privateKey;
+    /**
+     * @var PostBinding
+     */
+    private $postBinding;
+
+    public function __construct(
+        PostBinding $postBinding
+    ) {
+        $this->postBinding = $postBinding;
+
+        // Todo: make keys configurable
+        $this->publicKey = __DIR__ . '/../../../../../vendor/surfnet/stepup-saml-bundle/src/Resources/keys/development_publickey.cer';
+        $this->privateKey = __DIR__ . '/../../../../../vendor/surfnet/stepup-saml-bundle/src/Resources/keys/development_privatekey.pem';
+    }
+
+    /**
+     * @param string $nameId
+     * @return RedirectResponse
+     * @throws Exception
+     */
+    public function handleAuthnRequestToAdfs(string $nameId): string
+    {
+        $authnRequest = AuthnRequestFactory::createNewRequest($this->getServiceProvider(), $this->getIdentityProvider());
+
+        // Use emailaddress as subject
+        $authnRequest->setSubject($nameId);
+
+        // Set authnContextClassRef to force MFA
+        $authnRequest->setAuthenticationContextClassRef('http://schemas.microsoft.com/claims/multipleauthn');
+
+        // Build request query parameters.
+        $requestAsXml = $authnRequest->getUnsignedXML();
+        $encodedRequest = base64_encode(gzdeflate($requestAsXml));
+        $queryParams = [AuthnRequest::PARAMETER_REQUEST => $encodedRequest];
+
+        // Create redirect response.
+        $query = $this->signRequestQuery($queryParams);
+        return sprintf('%s?%s', $this->getIdentityProvider()->getSsoUrl(), $query);
```

Comment:
You might have investigated this, but wouldn't using: `\Surfnet\SamlBundle\SAML2\AuthnRequest::buildRequestQuery` or even `\SAML2\HTTPRedirect::getRedirectURL` be far more efficient? It seems to do the same stuff, but in the saml bundle/saml2 library.

Reply:
I shamefully copied the required code from the SPController.  I've changed the code to use the SAML2 methods instead and retroactively did this in the SPController.